File: app/utils/secret_manager.py
import logging
from app.utils.config import Settings

settings = Settings()

logger = logging.getLogger(__name__)


class SecretNotFoundError(Exception):
    """Raised when a secret is not found in Secret Manager."""

    pass


class SecretAccessError(Exception):
    """Raised when there's an error accessing a secret."""

    pass


class InvalidAPIKeyError(Exception):
    """
    Raised when an API key is invalid or missing.
    This is a permanent error that should not trigger Pub/Sub retries.
    """

    pass


def get_user_api_key(user_id: str, provider: str = "openai") -> str:
    """
    Fetches a user's API key from Google Cloud Secret Manager.

    Args:
        user_id: The user ID (customer_identifier) to fetch the key for
        provider: The API provider (default: "openai")

    Returns:
        The API key as a string

    Raises:
        SecretNotFoundError: If the secret doesn't exist
        SecretAccessError: If there's an error accessing the secret
    """
    from google.cloud import secretmanager
    from google.api_core import exceptions

    project_id = settings.gcp_project_id
    if not project_id:
        raise SecretAccessError("GCP project ID is not configured")

    secret_name = f"user-{user_id}-{provider}-key"
    resource_name = f"projects/{project_id}/secrets/{secret_name}/versions/latest"

    try:
        client = secretmanager.SecretManagerServiceClient()
        response = client.access_secret_version(request={"name": resource_name})
        api_key = response.payload.data.decode("UTF-8")
        return api_key
    except exceptions.NotFound:
        logger.error(f"Secret not found for user {user_id}")
        raise SecretNotFoundError(f"API key not found for user {user_id}")
    except Exception as e:
        logger.error(f"Error accessing secret for user {user_id}: {e}")
        raise SecretAccessError(f"Failed to access API key: {str(e)}")

File: apps/backend/internal/service/secret_manager.go
package service

import (
	"context"
	"fmt"

	"app/internal/config"

	secretmanager "cloud.google.com/go/secretmanager/apiv1"
	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
	"google.golang.org/api/option"
)

type SecretManagerService interface {
	StoreUserAPIKey(ctx context.Context, userID, provider, apiKey string) error
	GetUserAPIKey(ctx context.Context, userID, provider string) (string, error)
	DeleteUserAPIKey(ctx context.Context, userID, provider string) error
}

type secretManagerService struct {
	client    *secretmanager.Client
	projectID string
}

func NewSecretManagerService(ctx context.Context, cfg *config.Config) (SecretManagerService, error) {
	projectID := cfg.GetGCPProjectID()
	if projectID == "" {
		return nil, fmt.Errorf("GCP Project ID is not set for the current environment")
	}

	var opts []option.ClientOption
	// Note: Secret Manager requires a real GCP project even for local development.
	// Set GCP_PROJECT_ID_LOCAL to your local development GCP project ID.

	client, err := secretmanager.NewClient(ctx, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to create Secret Manager client: %w", err)
	}

	return &secretManagerService{
		client:    client,
		projectID: projectID,
	}, nil
}

func (s *secretManagerService) StoreUserAPIKey(ctx context.Context, userID, provider, apiKey string) error {
	secretName := fmt.Sprintf("user-%s-%s-key", userID, provider)
	secretPath := fmt.Sprintf("projects/%s/secrets/%s", s.projectID, secretName)

	secretExists := true
	_, err := s.client.GetSecret(ctx, &secretmanagerpb.GetSecretRequest{
		Name: secretPath,
	})
	if err != nil {
		secretExists = false
	}

	if !secretExists {
		createReq := &secretmanagerpb.CreateSecretRequest{
			Parent:   fmt.Sprintf("projects/%s", s.projectID),
			SecretId: secretName,
			Secret: &secretmanagerpb.Secret{
				Replication: &secretmanagerpb.Replication{
					Replication: &secretmanagerpb.Replication_Automatic_{
						Automatic: &secretmanagerpb.Replication_Automatic{},
					},
				},
			},
		}
		_, err := s.client.CreateSecret(ctx, createReq)
		if err != nil {
			return fmt.Errorf("failed to create secret: %w", err)
		}
	}

	addVersionReq := &secretmanagerpb.AddSecretVersionRequest{
		Parent: secretPath,
		Payload: &secretmanagerpb.SecretPayload{
			Data: []byte(apiKey),
		},
	}

	_, err = s.client.AddSecretVersion(ctx, addVersionReq)
	if err != nil {
		return fmt.Errorf("failed to add secret version: %w", err)
	}

	return nil
}

func (s *secretManagerService) GetUserAPIKey(ctx context.Context, userID, provider string) (string, error) {
	secretName := fmt.Sprintf("user-%s-%s-key", userID, provider)
	resourceName := fmt.Sprintf("projects/%s/secrets/%s/versions/latest", s.projectID, secretName)

	req := &secretmanagerpb.AccessSecretVersionRequest{
		Name: resourceName,
	}

	result, err := s.client.AccessSecretVersion(ctx, req)
	if err != nil {
		return "", fmt.Errorf("failed to access secret version: %w", err)
	}

	return string(result.Payload.Data), nil
}

func (s *secretManagerService) DeleteUserAPIKey(ctx context.Context, userID, provider string) error {
	secretName := fmt.Sprintf("user-%s-%s-key", userID, provider)
	secretPath := fmt.Sprintf("projects/%s/secrets/%s", s.projectID, secretName)

	req := &secretmanagerpb.DeleteSecretRequest{
		Name: secretPath,
	}

	err := s.client.DeleteSecret(ctx, req)
	if err != nil {
		return fmt.Errorf("failed to delete secret: %w", err)
	}

	return nil
}
